### PR TITLE
[Distributed] more restrictions on distributed funcs based on serialization findings

### PR DESCRIPTION
**Pull request by ktoso**
- We always intended to ban `inout`
- Varargs we can't support it seems, so ban them right away

Resolves rdar://85442872

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDistributed.cpp [hunk 1]
```diff
@@ -69,6 +69,24 @@ void swift::diagnoseDistributedFunctionInNonDistributedActorProtocol(
...
+  if (nominal->getInherited().empty()) {
+    SourceLoc fixItLoc = nominal->getBraces().Start;
+    diag.fixItInsert(fixItLoc, ": Codable");
+  } else {
+    ASTContext &ctx = nominal->getASTContext();
+    SourceLoc fixItLoc = nominal->getInherited().back().getSourceRange().End;
+    fixItLoc = Lexer::getLocForEndOfToken(ctx.SourceMgr, fixItLoc);
+    diag.fixItInsert(fixItLoc, ", Codable");
+  }
+}
```
Comment (ktoso) [inline lib/Sema/TypeCheckDistributed.cpp:88]: Similar to `addSendableFixIt` 👍

File: lib/Sema/TypeCheckDistributed.cpp [hunk 2]
```diff
@@ -119,28 +137,51 @@ bool swift::checkDistributedFunction(FuncDecl *func, bool diagnose) {
...
+      }
+      return true;
+    }
+
+    if (param->isInOut()) {
+      param->diagnose(
+          diag::distributed_actor_func_inout,
+          param->getName(),
+          func->getDescriptiveKind(), func->getName()
+      ).fixItRemove(param->getSpecifierLoc()); // FIXME(distributed): does not seem to cause the fixit...? Is the Loc invalid...?
```
Comment (ktoso) [inline lib/Sema/TypeCheckDistributed.cpp:157]: This is weird... should work, right? But it does not seem to cause the fixit (i verified using the `...}}{{12-14=}}` and nothing 🤔

Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:157]: @bnbarham maybe knows?

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDistributed.cpp (hunk 2 above)
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:157]: So it turns out this SpecifierLoc is `Invalid` 😢  Not sure why or how we'd fix it...

The func is:

```
(parameter "burger" apiName=inNOut type='String' interface type='String' inout)

  distributed func noInout(inNOut burger: inout String) { }
```

so I don't see what's "weird" about it hm...

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDistributed.cpp (hunk 2 above)
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp]: I found a workaround and asking around for help how to do it properly :)

Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp]: Oh boy, this is a bug with the SpecifierLoc -- it is _never_ correct currently.

I'll file a bug and work on it.

Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp]: https://bugs.swift.org/browse/SR-15486

**Review by xedin — APPROVED** [review]

**Pull request merged**